### ISSUE TITLE
.Net Core 5.0 BinaryFormatter + Self-contained = FileNotFoundException(mscorlib, v4)

### ISSUE THREAD

**Issue by scegg** [dotnet/runtime#50760]
### Description

Write a console app with BinaryFormatter called and compile that into self-contained, trimmode=link, publishtrimed, publishsinglefile . When run it under ubuntu arm64, an exception is thrown:
System.IO.FileNotFoundException: Could not load file or assembly 'mscorlib, Version=4.0.0.0

which is very similar with https://github.com/dotnet/runtime/issues/34494 but this is related to net 5.

Test with self-contained without other flag passed w/o problem.

### Configuration

.Net 5.0.201

BTW, I know that BinaryFormatter is obsoleted, lol.

**Comment by GrabYourPitchforks** [comment]
@marek-safar @eerhardt Is `BinaryFormatter` even supposed to work at all inside a trimmed app? Trying to figure out whether to route this issue to the trimming or serialization areas.

**Comment by ThomasGoulet73** [comment]
I could only reproduce this issue using trimming.

@scegg
You can use this workaround:
```
<Target Name="ConfigureTrimming" BeforeTargets="PrepareForILLink">
  <ItemGroup>
    <ManagedAssemblyToLink Condition="'%(Filename)' == 'mscorlib'">
      <IsTrimmable>false</IsTrimmable>
    </ManagedAssemblyToLink>
  </ItemGroup>
</Target>
```

**Comment by scegg** [comment]
> I could only reproduce this issue using trimming.
> 
> @scegg
> You can use this workaround:
> 
> ```
> <Target Name="ConfigureTrimming" BeforeTargets="PrepareForILLink">
>   <ItemGroup>
>     <ManagedAssemblyToLink Condition="'%(Filename)' == 'mscorlib'">
>       <IsTrimmable>false</IsTrimmable>
>     </ManagedAssemblyToLink>
>   </ItemGroup>
> </Target>
> ```

Thanks. Confirmed here also.

**Comment by vitek-karas** [comment]
`BinaryFormater` is simply not trim compatible - so we can say it's not supported with trimming turned on. Starting with .NET 6 Preview 3 the publish will produce a warning to this effect.

The main reason why it's not supported is that it can take type names in the deserialization payload and then tries to find the type by name in the application. Since trimming is a static analysis tool, it has no way to detect which types will be needed like that - and may remove some of them.

In this specific case, `mscorlib` is a facade assembly (there's nothing in it, just type forwarders), to optimize the size of the app, trimming may decide to remove the facade completely (and rewrite all type references to their real targets) - which is very likely what happens here.

**Comment by scegg** [comment]
> `BinaryFormater` is simply not trim compatible - so we can say it's not supported with trimming turned on. Starting with .NET 6 Preview 3 the publish will produce a warning to this effect.
> 
> The main reason why it's not supported is that it can take type names in the deserialization payload and then tries to find the type by name in the application. Since trimming is a static analysis tool, it has no way to detect which types will be needed like that - and may remove some of them.
> 
> In this specific case, `mscorlib` is a facade assembly (there's nothing in it, just type forwarders), to optimize the size of the app, trimming may decide to remove the facade completely (and rewrite all type references to their real targets) - which is very likely what happens here.

Here is my pseudocode:
```class Entity
{
xxx
}

class Program
{
  void main()
  {
  Entity value = new Entity {xxx = 123};
  ...
  BinaryFormatter.Serialize(value, memorystream);
  }
}
```

I only use serializer, not deserializer. The type to be serialized is used by direct code to set the value.

**Comment by vitek-karas** [comment]
Sorry - for serialization it's a known issue.
This is basically a duplicate of https://github.com/dotnet/runtime/issues/43132 which lead to https://github.com/mono/linker/issues/1569.

That said - because `BinaryFormatter` is obsolete - the priority of this issue is relatively low.

**Comment by eerhardt** [comment]
> That said - because BinaryFormatter is obsolete - the priority of this issue is relatively low.

I would say it is even lower than "relatively low". Basically we won't be fixing any issues with BinaryFormatter + trimming. For example, we had some code checked into main to make `DesigntimeLicenseContextSerializer` (which uses BinaryFormatter under the covers) to keep working in a trimmed app, but we backed it out with #50593. See the conversation in https://github.com/dotnet/runtime/pull/48527#discussion_r605216290.

Closing as we won't be fixing this.
